src/main/java/pl/javastart/jpaoptimalization/country/CountryService.java

```diff
@@ -16,4 +17,30 @@ public CountryService(CountryRepository countryRepository) {
     public List<Country> findAll() {
         return countryRepository.findAll();
     }
+
+    public List<CountryWithBiggestCity> findAllCountriesWithBiggestCity() {
+        return countryRepository.findAllCountriesWithBiggestCity();
+    }
+
+    public Map<String, List<LanguageWithUsage>> findAllCountriesWithLanguages() {
+        List<CountryWithLanguage> countriesWithLanguages = countryRepository.findAllCountriesWithLanguages();
+        Map<String, List<LanguageWithUsage>> countryWithLanguagesMap = new TreeMap<>();
+
+        for (CountryWithLanguage cl : countriesWithLanguages) {
+            String countryName = cl.getCountryName();
+            String language = cl.getLanguage();
+            Double percentage = cl.getPercentage();
+
+            LanguageWithUsage languageWithUsage = new LanguageWithUsage(language, percentage);
+
+            countryWithLanguagesMap
+                    .computeIfAbsent(countryName, k -> new ArrayList<>())
+                    .add(languageWithUsage);
+        }
+        return countryWithLanguagesMap;
+    }
```

Comment:
W klasie Country masz relację z CountryLanguage, więc wydaje się, że da się to zrobić zdecydowanie prościej. Przydało by się też uwzględnić kraje bez przypisanego żadnego języka w bazie, np. Antarctica

Reply:
Próbowałem wykonać to z pomocą projekcji jako dwa osobne interfejsy.

interface CountryWithLanguage {
String getCountryName();
LanguageAndUsage getLanguageAndUsage();
}

interface LanguageAndUsage {
String getLanguage();
Double getUsage();
}

Jednak Spring cały czas protestował.
/*
Invalid property '' of bean class [java.lang.String]: Could not find field for property during fallback access!
*/
Stackoverflow sugerował, żeby dodać dwustronną relację, @OneToMany @ManyToOne ale tego też nie udało mi się zaimplementować chyba ze względu na specyfikę identyfikatora języków, który nie wykorzystywał generatora ID, a miał dodatkową klasę odpowiadającą za identyfikator.